### PR TITLE
test: provide TransferData UTs

### PR DESCRIPTION
**Pull request by StefanCorneteanu**
## Description

The following changes are presented in this PR:

1. Refactored some printing statements to output a newline
2. WrongBlockSequenceNumberTest failed prematurely as the OTA state was not set to an expected value. In that regard, I renamed the current WrongBlockSequenceNumberTest to IncorrectOTAStateTest and redid WrongBlockSequenceNumberTest with a valid OTAState
3. The wrong frame was expected for the PositiveResponseTest; i changed it to the actual expected frame and, for coverage purposes, I added a call to getChecksums.
4. for both WrongBlockSequenceNumberTest and PositiveResponseTest I added a couple extra capture calls because the expected frame is not the first one sent
5. Added an unit test to cover the edge case when the expected block sequence number overflows the max uint8_t value (0xFF).
6. Added 2 new unit tests to cover the cases when the ota update status is PROCESSING_TRANSFER_COMPLETE, for both when the transfer is complete and when it fails
7. Added 3 new unit tests for when the processing for data transfer: one for the case when the processing fails, one for when the transfer starts and completes and one for when the transfer starts but isnt complete and one for the case when the transfer starts and completes
8. Despite the fact that the test cases above covered also calls to computeChecksum, it showed as not covered in the coverage report. For that case, I created a short UT

All tests passed and coverage is 100% (see ticket for screenshot)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Review by DirvaNIcolae — APPROVED** [review]

File: ota/transfer_data/utest/TransferDataTest.cpp [hunk 1]
```diff
@@ -64,19 +101,172 @@ TEST_F(TransferDataTest, WrongBlockSequenceNumberTest) {
...
+
+    // Set the OTA_UPDATE_STATUS_DID to PROCESSING_TRANSFER_COMPLETE to cover the data transfer fail section
+    FileManager::setDidValue(OTA_UPDATE_STATUS_DID, {PROCESSING_TRANSFER_COMPLETE}, kCanId, mockLogger, socket_);
+    // Using an extremely large frame so that MemoryManager::writeToAddress fails as availableMemory will fail
+    std::vector<uint8_t> frame_data (123456789,0x01);
+
+    testing::internal::CaptureStdout();
+    transfer_data->transferData(kCanId, frame_data);
+    std::string output = testing::internal::GetCapturedStdout();
+    EXPECT_NE(output.find("Data transfer failed at writting in memory."), std::string::npos);
```
Comment (AlexZahariaRandstad) [inline ota/transfer_data/utest/TransferDataTest.cpp:165]: Same problem related to captured output.

File: ota/transfer_data/utest/TransferDataTest.cpp [hunk 2]
```diff
@@ -64,19 +101,172 @@ TEST_F(TransferDataTest, WrongBlockSequenceNumberTest) {
...
+ */
+TEST_F(TransferDataTest, DataTransferComplete) {
+
+    // Set the OTA_UPDATE_STATUS_DID to PROCESSING_TRANSFER_COMPLETE to cover the data transfer complete section
+    FileManager::setDidValue(OTA_UPDATE_STATUS_DID, {PROCESSING_TRANSFER_COMPLETE}, kCanId, mockLogger, socket_);
+    std::vector<uint8_t> frame_data = {0x02, 0x36, 0x01, 0x02, 0x33};
+
+    testing::internal::CaptureStdout();
+    transfer_data->transferData(kCanId, frame_data);
+    std::string output = testing::internal::GetCapturedStdout();
```
Comment (AlexZahariaRandstad) [inline ota/transfer_data/utest/TransferDataTest.cpp:149]: Same problem related to captured output.

**Review by AlexZahariaRandstad — CHANGES_REQUESTED** [review]

File: ota/transfer_data/utest/TransferDataTest.cpp [hunk 3]
```diff
@@ -64,19 +101,172 @@ TEST_F(TransferDataTest, WrongBlockSequenceNumberTest) {
...
+    system(kStr_zipCreateCmd.c_str());
+    MemoryManager::writeToFile(vecU8_preprocessedData, kStr_zipFilePath, mockLogger);
+    std::string ecuPath;
+    FileManager::getEcuPath(0x11, ecuPath, 0, mockLogger, "0");
+
+    // Signal the start of processing data
+    testing::internal::CaptureStdout();
+    TransferData::processDataForTransfer(kCanId, vecU8_preprocessedData, socket_, mockLogger);
+    std::string output = testing::internal::GetCapturedStdout();
+    EXPECT_NE(output.find("Data processing for transfer started."), std::string::npos);
```
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply ota/transfer_data/utest/TransferDataTest.cpp:241]: Did not find an alternative way to test this behavior

File: ota/transfer_data/utest/TransferDataTest.cpp [hunk 4]
```diff
@@ -64,19 +101,172 @@ TEST_F(TransferDataTest, WrongBlockSequenceNumberTest) {
...
+    // Set OTA_UPDATE_STATUS_DID to WAIT_DOWNLOAD_COMPLETED, the state before starting to process data
+    FileManager::setDidValue(OTA_UPDATE_STATUS_DID, {WAIT_DOWNLOAD_COMPLETED}, MCU_ID, mockLogger, socket_);
+
+    // Create the zip file, write data to it, and get its path
+    system(kStr_zipCreateCmd.c_str());
+    MemoryManager::writeToFile(vecU8_preprocessedData, kStr_zipFilePath, mockLogger);
+    std::string ecuPath;
+    FileManager::getEcuPath(0x11, ecuPath, 0, mockLogger, "0");
+
+    // Signal the start of processing data
```
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply ota/transfer_data/utest/TransferDataTest.cpp:196]: Did not find an alternative way to test this behavior

File: ota/transfer_data/utest/TransferDataTest.cpp (hunk 1 above)
Comment (StefanCorneteanu) [reply ota/transfer_data/utest/TransferDataTest.cpp]: Changed testing method from output capturing to frame capturing

File: ota/transfer_data/utest/TransferDataTest.cpp (hunk 2 above)
Comment (StefanCorneteanu) [reply ota/transfer_data/utest/TransferDataTest.cpp]: Changed testing method from output capturing to frame capturing

**Review by AlexZahariaRandstad — APPROVED** [review]

File: ota/transfer_data/utest/TransferDataTest.cpp [hunk 5]
```diff
@@ -25,23 +32,51 @@ class TransferDataTest : public ::testing::Test {
...
     }
+    void SetUp(){
+        TransferData::expected_block_sequence_number = 0x01;
+    }
 };
 
+/* Test for computing the checksum */
+TEST_F(TransferDataTest, computeChecksumTest){
+    const std::vector<uint8_t> kVecU8_data {0x01, 0x02, 0x03};
+    const uint8_t checksum = TransferData::computeChecksum(kVecU8_data.data(), kVecU8_data.size());
```
Comment (VladVestemeanuRandstad) [inline ota/transfer_data/utest/TransferDataTest.cpp:51]: "u8Checksum" to align with new coding guidelines?

File: ota/transfer_data/utest/TransferDataTest.cpp [hunk 6]
```diff
@@ -25,23 +32,51 @@ class TransferDataTest : public ::testing::Test {
...
+
+    // Capture a garbage frame else it will impact the next tests
+    captured_frame->capture();
+}
+
+/* Test for IncorrectOTAState */
+TEST_F(TransferDataTest, IncorrectOTAState) {
+
+    // Do not set the OTA state to one of the expected states: WAIT_DOWNLOAD_COMPLETED, PROCESSING AND PROCESSING_TRANSFER_COMPLETE
+    std::vector<uint8_t> invalid_frame_data = {0x02, 0x36, 0x03, 0x02};
```
Comment (VladVestemeanuRandstad) [inline ota/transfer_data/utest/TransferDataTest.cpp:76]: u8InvalidFrameData

File: ota/transfer_data/utest/TransferDataTest.cpp [hunk 7]
```diff
@@ -64,19 +104,175 @@ TEST_F(TransferDataTest, WrongBlockSequenceNumberTest) {
...
+
+/* 
+ * Test for DataTransferComplete. 
+ * @warning Make sure that you ran create_sd_card.sh before running this test! 
+ */
+TEST_F(TransferDataTest, DataTransferComplete) {
+
+    // Set the OTA_UPDATE_STATUS_DID to PROCESSING_TRANSFER_COMPLETE to cover the data transfer complete section
+    FileManager::setDidValue(OTA_UPDATE_STATUS_DID, {PROCESSING_TRANSFER_COMPLETE}, kCanId, mockLogger, socket_);
+    std::vector<uint8_t> frame_data = {0x02, 0x36, 0x01, 0x02, 0x33};
```
Comment (VladVestemeanuRandstad) [inline ota/transfer_data/utest/TransferDataTest.cpp:132]: vecU8_FrameData
